### PR TITLE
Add dependency enforcer

### PR DESCRIPTION
**Pull request by bjorncs**
Compares all dependencies (including transitive) for all modules in Maven project.

Currently no classification of artifacts used for test classpath only. That will come later.

### CODE REVIEW

File: vespa-enforcer-extensions/src/test/java/com/yahoo/vespa/maven/plugin/enforcer/EnforceDependenciesAllProjectsTest.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+                Dependency.fromString("com.example:bar:2.3.4")));
+        Path outputFile = tempDir.resolve("allowed-dependencies.txt");
+        writeDependencySpec(outputFile, dependencies);
+        assertEquals(
+                Files.readString(Paths.get("src/test/resources/allowed-dependencies.txt")),
+                Files.readString(outputFile));
+
+    }
+
+}
```
Comment (baldersheim) [inline vespa-enforcer-extensions/src/test/java/com/yahoo/vespa/maven/plugin/enforcer/EnforceDependenciesAllProjectsTest.java:84]: Missing EOL at EOF

File: vespa-enforcer-extensions/src/test/java/com/yahoo/vespa/maven/plugin/enforcer/EnforceDependenciesAllProjectsTest.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+                                     () -> validateDependencies(dependencies, specFile));
+        String expectedErrorMessage =
+                """
+                The dependency enforcer failed:
+                Removed dependencies:
+                 - com.example:bar:2.3.4
+                Maven dependency validation failed. To update dependency spec run 'mvn enforcer:enforce -DdependencyEnforcer.writeSpec'""";
+        assertEquals(expectedErrorMessage, exception.getMessage());
+    }
+
```
Comment (baldersheim) [inline vespa-enforcer-extensions/src/test/java/com/yahoo/vespa/maven/plugin/enforcer/EnforceDependenciesAllProjectsTest.java:70]: Do you need to test that version has changed.
 Fx com.example:bar:2.3.4 vs com.example:bar:2.3.5 ?

**Review by baldersheim — APPROVED** [review]
Consider comments.

**Review by gjoranv — APPROVED** [review]
💯

File: vespa-enforcer-extensions/src/test/java/com/yahoo/vespa/maven/plugin/enforcer/EnforceDependenciesAllProjectsTest.java (hunk 2 above)
Comment (bjorncs) [reply vespa-enforcer-extensions/src/test/java/com/yahoo/vespa/maven/plugin/enforcer/EnforceDependenciesAllProjectsTest.java:70]: Different versions are handled as separate artifacts

**Pull request merged**